packages/ingest/recursion.js

```diff
@@ -1,84 +1,93 @@
- * @returns {Promise} the promise of an object that has the path is the key and
- *   list of files as values
+ * @param {string} segment - path segment
+ * @returns {boolean}
  */
-async function recursion(fn, originalPath, currentPath = null, position = 0) {
-  // build the recursion path object
-  const regex = /(\(.*?\))/g;
-  const rules = (originalPath || '/').split(regex).map((i) => i.replace(/\\\\/g, '\\'));
-  const map = rules.map((r) => (r.match(regex) !== null));
-
-  let files = [];
-  let path = currentPath;
-  if (!path) {
-    path = rules[position];
-  }
+const isRegexSegment = (segment) => segment.match(/(\([^\)]*\))/g) !== null;
 
-  log.info(`Listing ${path}`);
-
-  // get list of current path
-  const list = await fn(path);
-
-  // loop try what is returned
-  for (let ctr = 0; ctr < list.length; ctr += 1) {
-    const item = list[ctr];
+/**
+ * Insert leading and terminating slashes into the path string if not present
+ *
+ * @param {string} path - path string
+ * @returns {string} normalized path
+ */
+const normalizeWithSlashes = (path) => {
+  let output = path;
+  if (!path.startsWith('/')) output = `/${output}`;
+  if (!path.endsWith('/')) output = `${output}/`;
+  return output;
+};
 
-    let regexPath;
-    let textPath;
-    let newPath;
+/**
+ * Recur on directory, list all files, and recur into any further directories,
+ * as specified regex segments allow.
+ * Note that the list function will be called with a path argument that includes
+ * both leading and terminating slashes. List functions will need to be able to handle
+ * or remove leading and terminating slashes accordingly.
+ *
+ * @param {Function} fn - list function
+ * @param {string} currentPath - current path to list
+ * @param {Array<string>} segments - path segments
+ * @param {number} position - current position in the segment list
+ */
+async function recurOnDirectory(fn, currentPath, segments, position) {
+  // check if we have a filter regex segment (e.g. '(dir.*)')
+  const filterExpr = segments[position + 1] || '';
+  const isRegex = isRegexSegment(filterExpr);
 
-    // if directory is found, apply recursion logic
-    if (item.type === 'd' || item.type === 1) {
-      // first we check if the next segment of the provided path
-      // is a regex rule
-      const isRegex = map[position + 1];
+  const path = currentPath.replace(/\/+/g, '/');
+  const contents = fn(path);
+  let files = [];
 
-      // if it is the regex rule, we use the rule to
-      // decide whether to do more recursion
-      if (isRegex) {
-        regexPath = new RegExp(rules[position + 1]);
-      } else {
-        // if it is just a path, we use that segment and its children
-        textPath = rules[position + 1];
+  for (let ctr = 0; ctr < contents.length; ctr += 1) {
+    const item = contents[ctr];
+    if (['-', 0].includes(item.type)) {
+      if (!isRegex || item.name.match(new RegExp(filterExpr))) {
```

Comment:
Micro-optimization, but this condition is used here in both `if` blocks, so perhaps we can pull it out into a variable and re-use it in both?

`const itemPassesFiltering = !isRegex || item.name.match(new RegExp(filterExpr))`

Reply:
So what happens if an item is filtered by regex but does not match the filter expression? Basically we don't include that file in our outputs or recur into that directory, which is the desired behavior?

Reply:
In addition, as in an earlier comment, I would suggest using `RegExp.test` rather than `String.match` since you are not concerned with using the match result, but only on whether or not a match is made. That is, you are only interested in a boolean value, which is precisely what `test` is intended for.

Reply:
I'm seeing different results/failures when using `test` vs. `match`. I've yet to get it to work correctly.

Reply:
👍

Reply:
@markdboyd Yeah, we are trying to allow people to exclude certain directories. Abdelhak pointed out to me that this is desirable as sometimes they only get the root directory so to scan the provider's entire file system before filtering on `granuleIdExtraction` is suboptimal, to say the least.